src/workerd/jsg/modules-new-test.c++

```diff
@@ -0,0 +1,1889 @@
+
+    registry->attachToIsolate(js, compilationObserver);
+
+    js.tryCatch([&] {
+      ModuleRegistry::resolve(js, "file:///bar", "default"_kjc);
+      JSG_FAIL_REQUIRE(Error, "Should have failed");
+    }, [&](Value exception) {
+      auto str = kj::str(exception.getHandle(js));
+      KJ_ASSERT(str == "TypeError: Circular module dependency with synchronous require: "
+                       "file:///bar");
+    });
+  });
+}
+
+// ======================================================================================
+KJ_TEST("Resolution occurs relative to the referrer") {
+  ResolveObserver observer;
+  CompilationObserver compilationObserver;
+  ModuleRegistry::Builder registryBuilder(observer);
+
+  ModuleBundle::BundleBuilder builder;
+  builder.addSyntheticModule("foo/bar",
+      Module::newDataModuleHandler(kj::heapArray<kj::byte>(0)));
+  builder.addSyntheticModule("bar",
+      Module::newDataModuleHandler(kj::heapArray<kj::byte>(0)));
+
+  // The base URL of the referrer is file:///foo/ ... so in each of the
+  // following cases, the specifier should be resolved relative to that.
+  // For instance, 'bar' should resolve as file:///foo/bar, while '../bar'
+  // should resolve as file:///bar
+
+  auto bar = kj::str("export * as abc from 'bar';"           // file:///foo/bar
+                     "export * as def from './bar';"         // file:///foo/bar
+                     "export * as ghi from '../bar';"        // file:///bar
+                     "export * as jkl from '/bar';"          // file:///bar
+                     "export * as lmn from '../foo/bar';");  // file:///foo/bar
+  builder.addEsmModule("foo/", bar.slice(0, bar.size()).attach(kj::mv(bar)));
+
+  auto registry = registryBuilder.add(builder.finish()).finish();
+
+  PREAMBLE([&](Lock& js) {
+    registry->attachToIsolate(js, compilationObserver);
+
+    js.tryCatch([&] {
+      auto abc = ModuleRegistry::resolve(js, "file:///foo/", "abc"_kjc);
+      auto def = ModuleRegistry::resolve(js, "file:///foo/", "def"_kjc);
+      auto ghi = ModuleRegistry::resolve(js, "file:///foo/", "ghi"_kjc);
+      auto jkl = ModuleRegistry::resolve(js, "file:///foo/", "jkl"_kjc);
+      auto lmn = ModuleRegistry::resolve(js, "file:///foo/", "lmn"_kjc);
+
+      KJ_ASSERT(abc.strictEquals(def));
+      KJ_ASSERT(abc.strictEquals(lmn));
+      KJ_ASSERT(!abc.strictEquals(ghi));
+      KJ_ASSERT(ghi.strictEquals(jkl));
+    }, [&](Value exception) {
+      js.throwException(kj::mv(exception));
+    });
+  });
+}
+
+// ======================================================================================
+
+KJ_TEST("Building a bundle from a capnp description works") {
+
+  capnp::MallocMessageBuilder builder;
+  auto bundle = builder.initRoot<workerd::jsg::Bundle>();
+
+  auto modules = bundle.initModules(3);
+
+  auto str = kj::str("export default 1+1;");
+  auto wasm = kj::heapArray<kj::byte>({
```

Comment:
Does KJ have something to avoid heap allocation here? Not important within a test of course, but I would find it good to know about.

Reply:
No, this could likely just be a `kj::ArrayPtr` but it's just a test so didn't bother